README.md
# alpine

## Overview

Alpine Linux  is a security-oriented,  lightweight Linux distribution  based on
musl libc and busybox.

[alpinelinux.org](https://alpinelinux.org/)

## Description

Use this script to build your own base system.

We've included the last ca-certificates files  in the repository to ensure that
all of our images are accurates.

## Tags

Supported tags.

- 3.5
- 3.6
- 3.7
- 3.8
- 3.9
- 3.10
- 3.11
- 3.12
- 3.13
- 3.14
- 3.15
- 3.16, latest

## Setup

On Debian, Devuan and Ubuntu you need the following packages:

```bash
sudo apt-get -qq -y install curl
```

You also need to be in the docker group to use Docker.

```bash
sudo usermod -a -G docker ${USER}
```

Finally you need to login on Docker Hub.

```bash
docker login
```

## Usage

You first  need to  choose which dist  between 3.5, 3.6,  3.7, 3.8,  3.9, 3.10,
3.11, 3.12, 3.13, 3.14 or 3.15 you want (3.15 will be the 'latest' tag) and you
need to choose you user (or organization) name on Docker Hub.

Show help.

```bash
./build.sh -h
```

Build your own Alpine image (eg. 3.12).

```bash
./build.sh -d 3.14 -u vptech
```

Build your own Alpine image (eg. 3.13) and push it on the Docker Hub.

```bash
./build.sh -d 3.15 -u vptech -p
```

## Limitations

Only work on Debian, Devuan and Ubuntu.

## Development

Please read carefully [CONTRIBUTING.md](CONTRIBUTING.md) before making a merge request.

## Miscellaneous

```text
    ╚⊙ ⊙╝
  ╚═(███)═╝
 ╚═(███)═╝
╚═(███)═╝
 ╚═(███)═╝
  ╚═(███)═╝
   ╚═(███)═╝
```


CHANGELOG.md
# [1.3.0](https://git.vptech.eu/veepee/vptech/container-images/alpine/compare/1.2.3...1.3.0) (2022-07-22)


### Features

* add Alpine 3.16 ([e6ae8ba](https://git.vptech.eu/veepee/vptech/container-images/alpine/commit/e6ae8ba231ef63f289f4423680b10033394041c5))

## [1.2.3](https://git.vptech.eu/veepee/vptech/container-images/alpine/compare/1.2.2...1.2.3) (2022-03-27)


### Continuous Integration

* fix main branch issue ([a1f2b77](https://git.vptech.eu/veepee/vptech/container-images/alpine/commit/a1f2b778cb3dfcba22cc840b51f06f159eb1fad1))

## [1.2.2](https://git.vptech.eu/veepee/vptech/container-images/alpine/compare/1.2.1...1.2.2) (2022-03-27)


### Bug Fixes

* readme issues ([3cac59d](https://git.vptech.eu/veepee/vptech/container-images/alpine/commit/3cac59d036ca68418c96d2b191f5c8c414853297))


### Chores

* **deps:** update dependency veepee/vptech/sre/iac/ci-cd/shellcheck to v1.2.0 ([dc9391d](https://git.vptech.eu/veepee/vptech/container-images/alpine/commit/dc9391d0ced03533aa5f98680958107d332d2fde))


### Documentation

* change ownership ([23f9af8](https://git.vptech.eu/veepee/vptech/container-images/alpine/commit/23f9af8be46492a76106457c944236ec98342b11))

## [1.2.1](https://git.vptech.eu/veepee/vptech/sre/iac/docker-images/alpine/compare/1.2.0...1.2.1) (2021-12-13)


### Bug Fixes

* semantic changelog ([87bedb8](https://git.vptech.eu/veepee/vptech/sre/iac/docker-images/alpine/commit/87bedb8cb3e3ee3125067df6b0c766f1d6823916))


build.sh
#!/usr/bin/env bash
# shellcheck disable=SC1039
# shellcheck disable=SC2034

set -e

PATH='/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin'

arch='amd64'
version='1.4'

function usage()
{
    cat <<EOF

NAME:
   build.sh - Docker images' builder of Alpine.

USAGE:
   build.sh -d <dist>

OPTIONS:
   -h, --help           Show help

   -d, --dist           Choose Alpine distribution
                        eg: 3.10, 3.11, 3.12, 3.13, 3.14, 3.15, 3.16

   -t, --timezone       Choose your preferred timezone
                        default: Europe/Amsterdam

   -u, --user           Docker Hub username or organisation
                        default: $USER

   -p, --push           Docker Hub push
                        default: no

   -l, --latest         Force the "latest"
                        default: 3.16

   -v, --verbose        Verbose mode

   -V, --version        Show version

VERSION:
   docker-alpine version: ${version}

EOF
}

function docker_bootstrap()
{
    # variables
    cdn="http://dl-cdn.alpinelinux.org"
    ftp="${cdn}/alpine/v${distid}/releases/x86_64"
    rootfs="${ftp}/alpine-minirootfs-${distname}-x86_64.tar.gz"
    image="/tmp/image-${distname}-${arch}"

    echo '-- bootstrap' 1>&3

    if [ "$(id -u)" -ne 0 ]
    then
        sudo='sudo'
    fi

    # clean old image
    if [ -f "${image}.tar" ]
    then
        ${sudo} rm -f "${image}.tar"
    fi

    if [ -f "${image}.tar.gz" ]
    then
        ${sudo} rm -f "${image}.tar.gz"
    fi

    if [ -d "/tmp/image-${distname}-${arch}" ]
    then
        ${sudo} rm -fr "${image}"
    fi

    # download minimal image
    echo " * download ${image}" 1>&3
    ${sudo} curl \
            --location \
            --output "${image}.tar.gz" \
            --silent \
            "${rootfs}"
    ${sudo} mkdir "${image}"
    ${sudo} tar \
            -x \
            -C "${image}" \
            -f "${image}.tar.gz" \
            --numeric-owner \
            -z

    # create /etc/timezone
    echo ' * /etc/timezone' 1>&3
    cat <<EOF | \
        ${sudo} tee "${image}/etc/timezone"
${timezone}
EOF

    # create /etc/resolv.conf
    echo ' * /etc/resolv.conf' 1>&3
    cat <<EOF | \
        ${sudo} tee "${image}/etc/resolv.conf"
nameserver 8.8.4.4
nameserver 8.8.8.8
EOF

    # create /etc/apk/repositories
    echo ' * /etc/apk/repositories' 1>&3
    cat <<EOF | \
        ${sudo} tee "${image}/etc/apk/repositories"
http://mirror.veepee.tech/alpine/v${distid}/main
http://mirror.veepee.tech/alpine/v${distid}/community
EOF

    # upgrade
    echo ' * apk upgrade' 1>&3
    ${sudo} chroot "${image}" sh -c \
            "apk update  -q && \
             apk upgrade -q"

    # add ca-certificates
    echo ' * apk add ca-certificates' 1>&3
    ${sudo} chroot "${image}" sh -c \
            "apk add -q ca-certificates"

    # add dumb-init
    echo ' * apk add dumb-init' 1>&3
    ${sudo} chroot "${image}" sh -c \
            "apk add -q dumb-init"

    # clean
    echo ' * clean image' 1>&3
    ${sudo} chroot "${image}" sh -c \
            "rm -f /var/cache/apk/*"

    # create archive
    if [ -f "${image}.tar" ]
    then
        ${sudo} rm "${image}.tar"
    fi
    ${sudo} tar -C "${image}" -c -f "${image}.tar" --numeric-owner .


}

# create images from bootstrap archive
function docker_import()
{
    echo "-- docker import from ${image}" 1>&3
    docker import "${image}.tar" "${user}alpine:${distname}"
    docker run "${user}alpine:${distname}" \
           echo " * build ${user}alpine:${distname}" 1>&3
    docker tag "${user}alpine:${distname}" "${user}alpine:${distid}"
    docker run "${user}alpine:${distid}" \
           echo " * build ${user}alpine:${distid}" 1>&3

    if [ "${distname}" = "${latest}" ]
    then
        docker tag "${user}alpine:${distname}" "${user}alpine:latest"
        docker run "${user}alpine:latest" \
               echo " * build ${user}alpine:latest" 1>&3
    fi
}

# push image to docker hub
function docker_push()
{
    echo "-- docker push" 1>&3
    echo " * push ${user}alpine:${distname}" 1>&3
    docker push "${user}alpine:${distname}"
    echo " * push ${user}alpine:${distid}" 1>&3
    docker push "${user}alpine:${distid}"

    if [ "${distname}" = "${latest}"  ]
    then
        echo " * push ${user}alpine:latest" 1>&3
        docker push "${user}alpine:latest"
    fi
}

while getopts 'hd:t:u:plvV' OPTIONS
do
    case ${OPTIONS} in
        h)
            # -h / --help
            usage
            exit 0
            ;;
        d)
            # -d / --dist
            dist=${OPTARG}
            ;;
        t)
            # -t / --timezone
            timezone=${OPTARG}
            ;;
        u)
            # -u / --user
            user="${OPTARG}/"
            ;;
        p)
            # -p / --push
            push='true'
            ;;
        l)
            # -l / --latest
            latest=${OPTARG}
            ;;
        v)
            # -v / --verbose
            verbose='true'
            ;;
        V)
            # -v / --version
            echo "${version}"
            exit 0
            ;;
        *)
            usage
            exit 1
            ;;
    esac
done

if [ ! -x "$(command -v curl)" ]
then
    echo "Please install curl (see README.md)"
    exit 1
fi

# -d / --dist
if [ -n "${dist}" ]
then
    case ${dist} in
        3.5|3.5.3)
            distname='3.5.3'
            distid='3.5'
            ;;
        3.6|3.6.5)
            distname='3.6.5'
            distid='3.6'
            ;;
        3.7|3.7.3)
            distname='3.7.3'
            distid='3.7'
            ;;
        3.8|3.8.5)
            distname='3.8.5'
            distid='3.8'
            ;;
        3.9|3.9.6)
            distname='3.9.6'
            distid='3.9'
            ;;
        3.10|3.10.9)
            distname='3.10.9'
            distid='3.10'
            ;;
        3.11|3.11.13)
            distname='3.11.13'
            distid='3.11'
            ;;
        3.12|3.12.12)
            distname='3.12.12'
            distid='3.12'
            ;;
        3.13|3.12.11)
            distname='3.13.11'
            distid='3.13'
            ;;
        3.14|3.14.7)
            distname='3.14.7'
            distid='3.14'
            ;;
        3.15|3.15.5)
            distname='3.15.5'
            distid='3.15'
            ;;
        3.16|3.16.1)
            distname='3.16.1'
            distid='3.16'
            ;;
        *)
            usage
            exit 1
            ;;
    esac
else
    usage
    exit 1
fi

# -t / --timezone
if [ -z "${timezone}" ]
then
    timezone='Europe/Amsterdam'
fi

# -u / --user
if [ -z "${user}" ]
then
    user=${USER}
fi

# -l / --latest
if [ -z "${latest}" ]
then
    latest='3.16'
fi

# -v / --verbose
if [ -z "${verbose}" ]
then
    exec 3>&1
    exec 1>/dev/null
    exec 2>/dev/null
else
    exec 3>&1
fi

docker_bootstrap
docker_import

if [ -n "${push}" ]
then
    docker_push
fi
# EOF
